### PR TITLE
Write stream events that timeout to write to internal buffer in separate thread

### PR DESCRIPTION
**Pull request by dinujoh**
Write stream events that timeout to write to internal buffer in separate thread

 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by graytaylor0 — DISMISSED** [review]

File: data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java [hunk 1]
```diff
@@ -220,18 +229,19 @@ record = document.getFullDocument().toJson(JSON_WRITER_SETTINGS);
...
-                                    LOG.debug("Write to buffer for line {} to {}", (recordCount - recordFlushBatchSize), recordCount);
-                                    writeToBuffer(records, checkPointToken, recordCount);
-                                    lastLocalCheckpoint = checkPointToken;
-                                    lastLocalRecordCount = recordCount;
-                                    lastBufferWriteTime = System.currentTimeMillis();
-                                    bytesProcessedSummary.record(recordBytes.stream().mapToLong(Long::longValue).sum());
-                                    records.clear();
-                                    recordBytes.clear();
+
+                                lock.lock();
```
Comment (chenqi0805) [inline data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java:233]: Missing some context. Isn't this worker instance supposed to be owned by a single thread?

File: data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java [hunk 2]
```diff
@@ -304,13 +318,49 @@ private void writeToBuffer(final List<Event> records, final String checkPointTok
...
+            }
+
+            if (!sourceConfig.isAcknowledgmentsEnabled()) {
+                if (System.currentTimeMillis() - lastCheckpointTime >= checkPointIntervalInMs) {
+                    try {
+                        lock.lock();
+                        LOG.debug("Perform regular checkpoint for resume token {} at record count {}", lastLocalCheckpoint, lastLocalRecordCount);
+                        partitionCheckpoint.checkpoint(lastLocalCheckpoint, lastLocalRecordCount);
+                    } catch (Exception e) {
+                        LOG.error("Exception checkpointing the current state. New thread should start the stream from previous checkpoint.", e);
```
[the remark this answers is not in the document]
Comment (dinujoh) [reply data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java:357]: will update

File: data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java [hunk 3]
```diff
@@ -220,18 +229,19 @@ record = document.getFullDocument().toJson(JSON_WRITER_SETTINGS);
...
-                                    records.clear();
-                                    recordBytes.clear();
+
+                                lock.lock();
+                                try {
+                                    recordCount += 1;
+                                    checkPointToken = document.getResumeToken().toJson(JSON_WRITER_SETTINGS);
+
+                                    if ((recordCount % recordFlushBatchSize == 0) || (System.currentTimeMillis() - lastBufferWriteTime >= bufferWriteIntervalInMs)) {
+                                        writeToBuffer();
```
[the remark this answers is not in the document]
Comment (dinujoh) [reply data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java:239]: Main loop buffers record and writes when batch size or flush time is reached. The thread writes if main loop has not written for the time duration. The main thread will be blocked if there are no records/updates in stream and it has not flushed the data.

File: data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java [hunk 4]
```diff
@@ -119,13 +128,14 @@ public StreamWorker(final RecordBufferWriter recordBufferWriter,
         this.bytesReceivedSummary = pluginMetrics.summary(BYTES_RECEIVED);
         this.bytesProcessedSummary = pluginMetrics.summary(BYTES_PROCESSED);
         this.executorService = Executors.newSingleThreadExecutor(BackgroundThreadFactory.defaultExecutorThreadFactory("mongodb-stream-checkpoint"));
+        this.lock = new ReentrantLock();
         if (sourceConfig.isAcknowledgmentsEnabled()) {
             // starts acknowledgement monitoring thread
             streamAcknowledgementManager.init((Void) -> stop());
```
[the remark this answers is not in the document]
Comment (dinujoh) [reply data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java:134]: I will have a separate PR for handling streamAcknowledgmentManager. We will need to handle some edge cases.

File: data-prepper-plugins/mongodb/src/test/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorkerTest.java [hunk 5]
```diff
@@ -310,9 +310,9 @@ void test_processStream_checkPointIntervalSuccess() {
         verify(cursor).close();
         verify(cursor, times(4)).hasNext();
         verify(mockPartitionCheckpoint).getGlobalS3FolderCreationStatus(collection);
-        verify(mockPartitionCheckpoint, atLeast(1)).checkpoint(resumeToken3, 3);
+        //verify(mockPartitionCheckpoint, atLeast(1)).checkpoint(resumeToken3, 3);
```
[the remark this answers is not in the document]
Comment (dinujoh) [reply data-prepper-plugins/mongodb/src/test/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorkerTest.java:313]: will take a look

File: data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java (hunk 1 above)
Comment (dinujoh) [reply data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java:233]: We have main thread and new thread now. Main loop buffers record and writes when batch size or flush time is reached. The thread writes if main loop has not written for the time duration. The main thread will be blocked if there are no records/updates in stream and it has not flushed the data.

File: data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dinujoh) [reply data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java:135]: Added the change.

**Review by chenqi0805 — APPROVED** [review]

**Review by graytaylor0 — APPROVED** [review]

**Review by chenqi0805 — APPROVED** [review]
